hack/generator/pkg/codegen/pipeline_create_storage_types.go

```diff
@@ -0,0 +1,234 @@
+	}
+
+	storageRef := astmodel.MakeStoragePackageReference(localRef)
+	visitedName := astmodel.MakeTypeName(storageRef, name.Name())
+	return visitedName, nil
+}
+
+func (factory *StorageTypeFactory) visitObjectType(
+	_ *astmodel.TypeVisitor,
+	object *astmodel.ObjectType,
+	ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+	oc := vc.forObject(object)
+
+	var errs []error
+	properties := object.Properties()
+	for i, prop := range properties {
+		p, err := factory.makeStorageProperty(prop, oc)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			properties[i] = p
+		}
+	}
+
+	if len(errs) > 0 {
+		err := kerrors.NewAggregate(errs)
+		return nil, err
+	}
+
+	ot := astmodel.NewObjectType().WithProperties(properties...)
+	return astmodel.NewStorageType(*ot), nil
+}
+
+// makeStorageProperty applies a conversion to make a variant of the property for use when
+// serializing to storage
+func (factory *StorageTypeFactory) makeStorageProperty(
+	prop *astmodel.PropertyDefinition,
+	pc StorageTypesVisitorContext) (*astmodel.PropertyDefinition, error) {
+	for _, conv := range factory.propertyConversions {
+		p, err := conv(prop, pc.forProperty(prop))
+		if err != nil {
+			// Something went wrong, return the error
+			return nil, err
+		}
+		if p != nil {
+			// We have the conversion we need, return it promptly
+			return p, nil
+		}
+	}
+
+	return nil, fmt.Errorf("failed to find a conversion for property %v", prop.PropertyName())
+}
+
+// Preserve properties required by the KubernetesResource interface unchanged as they're always required
+func (factory *StorageTypeFactory) preserveKubernetesResourceStorageProperties(
+	prop *astmodel.PropertyDefinition,
+	_ StorageTypesVisitorContext) (*astmodel.PropertyDefinition, error) {
+	if prop.HasName(astmodel.AzureNameProperty) ||
+		prop.HasName(astmodel.OwnerProperty) {
+		// Keep these unchanged
+		return prop, nil
+	}
+
+	// No opinion, defer to another conversion
+	return nil, nil
+}
+
+func (factory *StorageTypeFactory) convertPropertiesForStorage(
+	prop *astmodel.PropertyDefinition,
+	ctx StorageTypesVisitorContext) (*astmodel.PropertyDefinition, error) {
+	propertyType, err := factory.visitor.Visit(prop.PropertyType(), prop)
+	if err != nil {
+		return nil, err
+	}
+
+	p := prop.WithType(propertyType).
+		MakeOptional().
+		WithoutValidation().
+		WithDescription("")
+
+	return p, nil
+}
+
+func (factory *StorageTypeFactory) visitArmType(
+	_ *astmodel.TypeVisitor,
+	it *astmodel.ArmType,
+	_ interface{}) (astmodel.Type, error) {
+	// We don't want to do anything with ARM types
+	return it, nil
+}
+
+func descriptionForStorageVariant(definition astmodel.TypeDefinition) []string {
+	pkg := definition.Name().PackageReference.PackageName()
+
+	result := []string{
+		fmt.Sprintf("Storage version of %v.%v", pkg, definition.Name().Name()),
+	}
+	result = append(result, definition.Description()...)
+
+	return result
+}
+
+type StorageTypesVisitorContext struct {
```

Comment:
Might be worth leaving a comment on this context type? It's not clear to me what exactly it's actually for. I tried to follow its usage to figure out what it is doing but as far as I can tell:
1. There is a single call to `forProperty`, which doesn't actually do anything because neither of the conversion methods actually use the provided context.
2. There is a single call to `forObject` which is passed to `makeStorageProperty` which then has `forProp` called on it and is passed to the conversion method, but see item 1 above as that value isn't used, which suggests the result of `forObject` _also_ actually isn't used?
3. There's a use of `vc.property != nil` but based on 1 and 2 above I can't see how that's getting hit ever.

I must be missing something but not sure what, and definitely not sure I follow exactly what this is tracking and what its purpose is.

Reply:
When visiting a `TypeName` different actions need to be taken, depending on what is being named. 

Currently, we only have one rule: An `EnumType` needs to be converted to the underlying base type, but only if the original `TypeName` is part of a property definition.

I'm expecting additional rules will be added as I work through milestones for storage types.